### ISSUE TITLE
upower support

### ISSUE THREAD

**Issue by z3ntu** [openrazer/openrazer#87]
> UPower is an abstraction for enumerating power devices, listening to device events and querying history and statistics.

https://upower.freedesktop.org/

(for mice (or other devices) with a built-in battery)

**Comment by terricain** [comment]
This'll be good but its going to be low on the priority list

**Comment by z3ntu** [comment]
Actually this might really be possible. When doing something with /sys/bus/hid/devices/<gamepad>/power_supply/ , see https://github.com/atar-axis/xpadneo/issues/21 the battery should be registered

**Comment by Aasikki** [comment]
This would be really nice to have. It's annoying to have to quess my mouses battery level.

**Comment by babbaj** [comment]
implemented this in https://github.com/openrazer/openrazer/pull/2160 with a power_supply. The xpadneo code was a very helpful reference.

**Comment by rashidmya** [comment]
Opened #2868 to implement this.

It adds a small generic `power_supply` helper in `razercommon` and wires it into the mouse and keyboard drivers, so wireless Razer mice/keyboards show up in UPower and the desktop battery tray (KDE "Power & Battery", GNOME "Power"), the way `hid-logitech-hidpp` surfaces a wireless mouse.

It's built to address the review feedback that the earlier attempt (#2160) was blocked on:
- the battery is refreshed on an async worker — `get_property` never blocks, it just reads a cache;
- `power_supply_changed()` is dedup'd (fires only on a real change);
- a sleeping/absent wireless device is reported as `present=0` (so the tray hides it instead of showing a false 0%), using a quiet query variant so it doesn't spam dmesg.

Caveat, stated in the PR too: I could only runtime-test the shared helper live on a wireless headset built on the same helper (separate branch). The mouse/keyboard paths are correct-by-construction — the existing per-PID battery query is reused unchanged and registration mirrors the exact set of devices that already expose `charge_level` — but I don't have a Razer wireless mouse/keyboard to verify them, so testing by anyone with the hardware would be very welcome.

**Comment by rashidmya** [comment]
Follow-up on device coverage: with #2868 handling wireless mice and keyboards, the wireless **BlackShark V3 / V3 Pro headset** battery is now wired to the same generic `power_supply` helper too. It's part of the V3 headset PR #2794 (added via mehmetbayoglu/openrazer#2), since that headset driver isn't in master yet and so rides with the V3 work rather than with #2868.

Unlike the mouse/keyboard paths, the headset one is **verified live**: it shows in UPower with the correct percentage and state, and reads 100% at full charge.

So once #2868 and #2794 both land, upower/tray battery support will cover all three wireless device families — mice, keyboards, and headsets — through one shared helper.
